src/hooks/useColumnWidths.ts

```diff
@@ -17,120 +17,96 @@ export function useColumnWidths<R, SR>(
   setMeasuredColumnWidths: StateSetter<ReadonlyMap<string, number>>,
   onColumnResize: DataGridProps<R, SR>['onColumnResize']
 ) {
-  const [autoResizeColumn, setAutoResizeColumn] = useState<CalculatedColumn<R, SR> | null>(null);
   const prevGridWidthRef = useRef(gridWidth);
   const columnsCanFlex: boolean = columns.length === viewportColumns.length;
   // Allow columns to flex again when...
   const ignorePreviouslyMeasuredColumns: boolean =
-    // there is enough space for columns to flex and...
-    columnsCanFlex &&
-    // the grid has resized, or a column is being `max-content`-sized.
-    (gridWidth !== prevGridWidthRef.current || autoResizeColumn !== null);
-
-  const columnsToMeasure = viewportColumns.filter(
-    ({ key, width }) =>
-      !resizedColumnWidths.has(key) &&
-      (ignorePreviouslyMeasuredColumns || !measuredColumnWidths.has(key)) &&
+    // there is enough space for columns to flex and the grid was resized
+    columnsCanFlex && gridWidth !== prevGridWidthRef.current;
+  const newTemplateColumns = [...templateColumns];
+  const columnsToMeasure: string[] = [];
+
+  for (const { key, idx, width } of viewportColumns) {
+    if (
       typeof width === 'string' &&
-      key !== autoResizeColumn?.key
-  );
+      (ignorePreviouslyMeasuredColumns || !measuredColumnWidths.has(key)) &&
+      !resizedColumnWidths.has(key)
+    ) {
+      newTemplateColumns[idx] = width;
+      columnsToMeasure.push(key);
+    }
+  }
+
+  const gridTemplateColumns = newTemplateColumns.join(' ');
 
   useLayoutEffect(() => {
     prevGridWidthRef.current = gridWidth;
+    updateMeasuredWidths(columnsToMeasure);
+  });
 
+  function updateMeasuredWidths(columnsToMeasure: readonly string[]) {
     if (columnsToMeasure.length === 0) return;
 
     setMeasuredColumnWidths((measuredColumnWidths) => {
       const newMeasuredColumnWidths = new Map(measuredColumnWidths);
       let hasChanges = false;
 
-      for (const { key } of columnsToMeasure) {
+      for (const key of columnsToMeasure) {
         const measuredWidth = measureColumnWidth(gridRef, key);
         hasChanges ||= measuredWidth !== measuredColumnWidths.get(key);
-        newMeasuredColumnWidths.set(key, measuredWidth);
+        if (measuredWidth === undefined) {
```

Comment:
When can it be undefined?

Reply:
Columns that are outside the viewport.
Previously it was resize -> remove columns from state -> re-render -> re-measure
Now it's resize -> re-measure -> update state with new size, delete previous measurement otherwise